### PR TITLE
[SPARK-38908][SQL] Provide query context in runtime error of Casting from String to Number/Date/Timestamp/Boolean

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Provide query context in runtime error of Casting from String to Number/Date/Timestamp/Boolean.
Casting Double/Float to Timestamp shares the same error method as casting String to Timestamp, so this PR also provides query context in its error.

### Why are the changes needed?

Provide SQL query context of runtime errors to users, so that they can understand it better.

### Does this PR introduce _any_ user-facing change?

Yes, improve the runtime error message of Casting from String to Number/Date/Timestamp/Boolean

### How was this patch tested?

UT

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/ansi/cast.sql.out [hunk 1]
```diff
@@ -448,6 +496,9 @@ struct<>
 -- !query output
 java.lang.NumberFormatException
 invalid input syntax for type numeric: '1中文'. To return NULL instead, use 'try_cast'. If necessary set spark.sql.ansi.enabled to false to bypass this error.
+== SQL(line 1, position 7) ==
+select cast('1中文' as tinyint)
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/results/ansi/cast.sql.out:500]: It's minor but seems the stop index is incorrect if the expression contains unicode.

File: sql/core/src/test/resources/sql-tests/results/ansi/cast.sql.out [hunk 2]
```diff
@@ -448,6 +496,9 @@ struct<>
...
+       ^^^^^^^^^^^^^^^^^^^^^^
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/results/ansi/cast.sql.out:501]: It's minor but seems the stop index is incorrect if the expression contains unicode.

**Comment by gengliangwang** [comment]
cc @entong as well

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Tests passed in https://github.com/gengliangwang/spark/actions/runs/2171731195
Merging to master/3.3.

**Pull request closed without merging**